### PR TITLE
net/gnrc/network_layer: remove timex_t

### PR DESCRIPTION
**Pull request by cgundogan**
This PR replaces `timex_t` (and corresponding functions) with `uint32_t` in the `net/gnrc/network_layer` folder.
Rationale: `timex_t` was used by vtimer. Now that all vtimer uses in the network_layer folder were replaced by xtimer calls, it makes no sense to still keep `timex_t` here.

_EDIT:_ (sizes for the `gnrc_networking` example)
```
make info-buildsizes-diff OLDBIN=master-bin NEWBIN=timex-bin

text    data    bss     dec     BOARD/BINDIRBASE

ERR     ERR     ERR     ERR     airfy-beacon
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-216    0       -8      -224    arduino-due
55088   224     20392   75704   master-bin
54872   224     20384   75480   timex-bin

-2036   0       -8      -2044   arduino-mega2560
80514   9450    14073   104037  master-bin
78478   9450    14065   101993  timex-bin

-280    0       0       -280    avsextrem
128512  2332    95961   226805  master-bin
128232  2332    95961   226525  timex-bin

-216    0       -8      -224    cc2538dk
54852   220     20368   75440   master-bin
54636   220     20360   75216   timex-bin

ERR     ERR     ERR     ERR     chronos
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-216    0       -8      -224    ek-lm4f120xl
54740   220     20360   75320   master-bin
54524   220     20352   75096   timex-bin

-208    0       -8      -216    f4vi1
54804   224     20360   75388   master-bin
54596   224     20352   75172   timex-bin

-216    0       -8      -224    fox
55344   220     20488   76052   master-bin
55128   220     20480   75828   timex-bin

-208    0       -8      -216    frdm-k64f
56780   1248    20360   78388   master-bin
56572   1248    20352   78172   timex-bin

-216    0       -8      -224    iotlab-m3
70196   220     23432   93848   master-bin
69980   220     23424   93624   timex-bin

-216    0       -8      -224    limifrog-v1
55996   220     20496   76712   master-bin
55780   220     20488   76488   timex-bin

-216    0       -8      -224    mbed_lpc1768
54512   220     20368   75100   master-bin
54296   220     20360   74876   timex-bin

ERR     ERR     ERR     ERR     msb-430
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     msb-430h
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-408    0       0       -408    msba2
155816  2332    95961   254109  master-bin
155408  2332    95961   253701  timex-bin

-216    0       -8      -224    msbiot
55236   224     20376   75836   master-bin
55020   224     20368   75612   timex-bin

-216    0       -8      -224    mulle
74968   1284    23712   99964   master-bin
74752   1284    23704   99740   timex-bin

-1600   0       -32     -1632   native
189638  624     114200  304462  master-bin
188038  624     114168  302830  timex-bin

ERR     ERR     ERR     ERR     nrf51dongle
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     nrf6310
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-304    0       -8      -312    nucleo-f091
57664   220     20368   78252   master-bin
57360   220     20360   77940   timex-bin

-208    0       -8      -216    nucleo-f303
54892   220     20376   75488   master-bin
54684   220     20368   75272   timex-bin

ERR     ERR     ERR     ERR     nucleo-f334
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-216    0       -8      -224    nucleo-l1
55924   220     20488   76632   master-bin
55708   220     20480   76408   timex-bin

-216    0       -8      -224    openmote
54760   220     20368   75348   master-bin
54544   220     20360   75124   timex-bin

-216    0       -8      -224    pba-d-01-kw2x
72020   1248    23784   97052   master-bin
71804   1248    23776   96828   timex-bin

ERR     ERR     ERR     ERR     pca10000
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     pca10005
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-280    0       0       -280    pttu
128720  2332    95961   227013  master-bin
128440  2332    95961   226733  timex-bin

-216    0       -8      -224    remote
55436   220     20368   76024   master-bin
55220   220     20360   75800   timex-bin

-304    0       -8      -312    saml21-xpro
57852   220     20488   78560   master-bin
57548   220     20480   78248   timex-bin

-336    0       -8      -344    samr21-xpro
74880   220     23448   98548   master-bin
74544   220     23440   98204   timex-bin

-216    0       -8      -224    slwstk6220a
54480   220     20488   75188   master-bin
54264   220     20480   74964   timex-bin

ERR     ERR     ERR     ERR     spark-core
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     stm32f0discovery
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-208    0       -8      -216    stm32f3discovery
54900   220     20376   75496   master-bin
54692   220     20368   75280   timex-bin

-216    0       -8      -224    stm32f4discovery
55084   224     20368   75676   master-bin
54868   224     20360   75452   timex-bin

ERR     ERR     ERR     ERR     telosb
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

-216    0       -8      -224    udoo
55088   224     20392   75704   master-bin
54872   224     20384   75480   timex-bin

ERR     ERR     ERR     ERR     weio
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     wsn430-v1_3b
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     wsn430-v1_4
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     yunjia-nrf51822
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin

ERR     ERR     ERR     ERR     z1
ERR     ERR     ERR     ERR     master-bin
ERR     ERR     ERR     ERR     timex-bin
```

### CODE REVIEW

**Comment by OlegHahm** [comment]
Have you checked how much bytes this saves?

**Comment by cgundogan** [comment]
@OlegHahm thanks for the hint. I updated the PR message. Looks fruitful.

**Comment by OlegHahm** [comment]
nice

File: sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c [hunk 1]
```diff
@@ -104,8 +103,8 @@ void gnrc_ndp_internal_set_state(gnrc_ipv6_nc_t *nc_entry, uint8_t state)
         case GNRC_IPV6_NC_STATE_REACHABLE:
             ipv6_iface = gnrc_ipv6_netif_get(nc_entry->iface);
             DEBUG("REACHABLE (reachable time = %" PRIu32 ".%06" PRIu32 ")\n",
-                  ipv6_iface->reach_time.seconds,
-                  ipv6_iface->reach_time.microseconds);
+                  ipv6_iface->reach_time / SEC_IN_USEC,
+                  ipv6_iface->reach_time % SEC_IN_USEC);
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: Just print the time in micro seconds.

File: sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c [hunk 2]
```diff
@@ -126,16 +125,14 @@ void gnrc_ndp_internal_set_state(gnrc_ipv6_nc_t *nc_entry, uint8_t state)
             nc_entry->probes_remaining = GNRC_NDP_MAX_UC_NBR_SOL_NUMOF;
             DEBUG("PROBE (probe with %" PRIu8 " unicast NS every %" PRIu32
                   ".%06" PRIu32 " seconds)\n", nc_entry->probes_remaining,
-                  ipv6_iface->retrans_timer.seconds,
-                  ipv6_iface->retrans_timer.microseconds);
+                  ipv6_iface->retrans_timer / SEC_IN_USEC,
+                  ipv6_iface->retrans_timer % SEC_IN_USEC);
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: Just print the time in microseconds

**Comment by authmillenon** [comment]
Please remove `timex` as dependency for `gnrc_ndp` in `Makefile.dep`

File: sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c [hunk 3]
```diff
@@ -92,7 +91,7 @@ ipv6_addr_t *gnrc_ndp_internal_default_router(void)
 void gnrc_ndp_internal_set_state(gnrc_ipv6_nc_t *nc_entry, uint8_t state)
 {
     gnrc_ipv6_netif_t *ipv6_iface;
-    timex_t t = { GNRC_NDP_FIRST_PROBE_DELAY, 0 };
+    uint32_t t = GNRC_NDP_FIRST_PROBE_DELAY * SEC_IN_USEC;
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/internal/gnrc_ndp_internal.c]: Can you please put this change in its own commit.

**Comment by authmillenon** [comment]
Tested so far:

- [ ] reachability time (`gnrc_ipv6_netif_t::reach_time`)
- [ ] retransmission timer (`gnrc_ipv6_netif_t::retrans_timer`)
- [ ] [probe delay timer](https://github.com/RIOT-OS/RIOT/pull/4216/files#diff-eb98861cc04f206d3376ec33720dfcc7R94)
- [ ] `_reschedule_rtr_sol()`
- [ ] ABR entry timeout.

**Comment by cgundogan** [comment]
addressed comments of @authmillenon

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c [hunk 4]
```diff
@@ -541,8 +540,7 @@ void gnrc_ndp_rtr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt, ipv6_hdr_t
     }
     /* set retransmission timer from message */
     if (rtr_adv->retrans_timer.u32 != 0) {
-        if_entry->retrans_timer = timex_set(0, byteorder_ntohl(rtr_adv->retrans_timer));
-        timex_normalize(&if_entry->retrans_timer);
+        if_entry->retrans_timer = byteorder_ntohl(rtr_adv->retrans_timer);
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: This is wrong. The [retransmission timer](https://tools.ietf.org/html/rfc4861#page-21) in a RA is send in milliseconds, while it is stored (and used as) a microsecond value in `gnrc_ipv6_netif_t::retrans_timer`. Can you maybe fix that in an a follow-up PR (+ hot-fix to the release branch)

File: sys/include/net/gnrc/ipv6/netif.h [hunk 5]
```diff
@@ -331,14 +331,14 @@ typedef struct {
...
-    timex_t reach_time;
+    uint32_t reach_time;
 
     /**
      * @brief   Time between retransmissions of neighbor solicitations to a
      *          neighbor.
      *          The default value is @ref GNRC_NDP_RETRANS_TIMER.
      */
-    timex_t retrans_timer;
+    uint32_t retrans_timer;
```
Comment (authmillenon) [inline sys/include/net/gnrc/ipv6/netif.h]: Both `reach_time` and `retrans_timer` need to be of `uint64_t` or handled as milliseconds altogether. Both values are [32-bit millisecond values](https://tools.ietf.org/html/rfc4861#page-21) in the RFC, but used as microseconds here.

Comment (authmillenon) [inline sys/include/net/gnrc/ipv6/netif.h]: Forget what I said. `reach_time` mustn't be greater than 1 h (well in 32-bit microseconds) and a `retrans_timer` greater than `reach_time` doesn't make much sense IMHO.

File: sys/net/gnrc/network_layer/ndp/gnrc_ndp.c (hunk 4 above)
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ndp/gnrc_ndp.c]: I'll do one, when I find some time… might be a little bit more complex.

**Comment by authmillenon** [comment]
ACK. Please squash.

**Comment by authmillenon** [comment]
Sorry, already squashed. You may merge when travis is in approval of your concoction :D
